[
    {
        "content": "<p>Hey guys, I am using the latest mathlib4 commit for a project and have the following two imports: </p>\n<div class=\"codehilite\" data-code-language=\"Lean\"><pre><span></span><code><span class=\"kn\">import</span> <span class=\"n\">Mathlib.Analysis.SpecialFunctions.Log.Base</span>\n<span class=\"kn\">import</span> <span class=\"n\">Mathlib.Data.Zmod.Basic</span>\n</code></pre></div>\n<p>Individually, each of these imports work fine. But when put together as above, it gives the following error: </p>\n<div class=\"codehilite\" data-code-language=\"Lean\"><pre><span></span><code><span class=\"kn\">import</span> <span class=\"n\">Mathlib.Data.Zmod.Basic</span> <span class=\"n\">failed</span><span class=\"o\">,</span> <span class=\"n\">environment</span> <span class=\"n\">already</span> <span class=\"n\">contains</span> <span class=\"bp\">'</span><span class=\"n\">Int.castAddHom._at.ZMod.lift._spec_1._cstage2'</span> <span class=\"k\">from</span> <span class=\"n\">Mathlib.Data.ZMod.Basic</span>\n</code></pre></div>\n<p>I am not sure how to parse this, perhaps there is some circular import happening somewhere? Some help would be appreciated, thank you!</p>",
        "id": 384574621,
        "sender_full_name": "Rahul Saha",
        "timestamp": 1691956439
    },
    {
        "content": "<p>Ah we've had this a few days ago! It's <code>Mathlib.Data.ZMod.Basic</code>with a capital <code>M</code>.</p>",
        "id": 384574832,
        "sender_full_name": "Anatole Dedecker",
        "timestamp": 1691956502
    },
    {
        "content": "<p>Let me guess, you're on MacOS?</p>",
        "id": 384574929,
        "sender_full_name": "Anatole Dedecker",
        "timestamp": 1691956533
    },
    {
        "content": "<p>yep MacOS</p>",
        "id": 384574954,
        "sender_full_name": "Rahul Saha",
        "timestamp": 1691956544
    },
    {
        "content": "<p>and thanks that worked!</p>",
        "id": 384575103,
        "sender_full_name": "Rahul Saha",
        "timestamp": 1691956592
    },
    {
        "content": "<p>An explanation of what goes wrong: <a href=\"#narrow/stream/287929-mathlib4/topic/EReal.20instance/near/383550857\">https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/EReal.20instance/near/383550857</a></p>",
        "id": 384575147,
        "sender_full_name": "Anatole Dedecker",
        "timestamp": 1691956615
    },
    {
        "content": "<p>Basically: you import the <code>Zmod</code> file (which does not exist), but MacOS is \"\"\"nice\"\"\" enough to give the <code>ZMod</code> file to Lean instead. But the problem is that another file imports <code>ZMod</code>, and Lean has no idea that these files are actually the same!</p>",
        "id": 384575334,
        "sender_full_name": "Anatole Dedecker",
        "timestamp": 1691956693
    },
    {
        "content": "<p>oh wow, that's quite silly behavior from the os, thanks for the explanation.</p>",
        "id": 384575437,
        "sender_full_name": "Rahul Saha",
        "timestamp": 1691956749
    },
    {
        "content": "<p>AFAIR, Windows has case-insensitive filesystem too.</p>",
        "id": 384580959,
        "sender_full_name": "Yury G. Kudryashov",
        "timestamp": 1691959049
    },
    {
        "content": "<p>I think it's case-insensitive in a different way! Read the extensive discussion in the linked thread :-)</p>",
        "id": 384594234,
        "sender_full_name": "Kevin Buzzard",
        "timestamp": 1691965950
    }
]